### ISSUE TITLE
chef resources should support umask attribute

### ISSUE THREAD

**Issue by foobarbam** [chef/chef#3741]
If chef_gem is installed using chef-client by a user with restrictive umask such as 0077.  chef_gem resources installs files with unreadable permissions, prohibiting use of chef code by other non-root users altogether.

Please facilitate umask attribute to chef_gem

``` shell
[root@enterprise ~]# umask  # chef-client runs like this
0077
[root@enterprise ~]# exit
logout
[kirk@enterprise ~]$ sudo ls -l /opt/chef/embedded/lib/ruby/gems/2.1.0/specifications/di-ruby-lvm-attrib-0.0.20.gemspec
-rw------- 1 root root 814 Jul 30 19:53 /opt/chef/embedded/lib/ruby/gems/2.1.0/specifications/di-ruby-lvm-attrib-0.0.20.gemspec
[kirk@enterprise ~]$ knife status
/opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/stub_specification.rb:71:in `initialize': Permission denied @ rb_sysopen - /opt/chef/embedded/lib/ruby/gems/2.1.0/specifications/di-ruby-lvm-attrib-0.0.20.gemspec (Errno::EACCES)
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/stub_specification.rb:71:in `open'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/stub_specification.rb:71:in `data'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/stub_specification.rb:167:in `valid?'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:733:in `block in each_stub'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:725:in `block (2 levels) in each_gemspec'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:724:in `each'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:724:in `block in each_gemspec'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:723:in `each'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:723:in `each_gemspec'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:731:in `each_stub'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:750:in `stubs'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/dependency.rb:271:in `matching_specs'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/dependency.rb:293:in `to_specs'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/dependency.rb:322:in `to_spec'
    from /opt/chef/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_gem.rb:58:in `gem'
    from /usr/bin/knife:4:in `<main>'
```

Aha! Link: https://chef.aha.io/features/SH-118

**Comment by lamont-granquist** [comment]
This is most likely a larger problem that we have with many resources, and the chef_gem resource is only one of them.  We most likely need a universal umask attribute which will set the umask before the resource converges and then restore it afterwards.

Setting the umask to 077 is somewhat common in Enterprise-ey environments with stringent security requirements.

**Comment by ranjib** [comment]
@foobarbam i dont think we'll be dealing with this any time soon, due to the wide range of resources this affects, along with the intended use case and how many users actually need this. Instead of supporting this on per resource basis, it might be easier to set the intended umask value using File#umask() from a recipe or library .. you can also use event handlers to do it before actual convergence begins. Albeit you have to deal with the consequences, but this also means you can selectively set this for the chef_gem resource.

**Comment by lamont-granquist** [comment]
We do support a config.rb setting for umask so that in 077 environments chef-client itself can be switched to run with 022 which will fix all these problems (maybe violating the letter of the 077 law, but that law is intended to apply to user-created-files by default and these kinds of exceptions with system processes are going to be common).

**Comment by foobarbam** [comment]
Good workarounds being suggested, thanks guys

**Comment by lamont-granquist** [comment]
Since there's a workaround dropping the priority of this.

**Comment by lamont-granquist** [comment]
Although I don't think this would be as difficult as @ranjib suggests, we could implement it on Chef::Resource similarly to not_if/only_if and other common properties and wrap the whole execution of the provider in a umask set/restore block.  Should be fairly straight-forwards code.

**Comment by mfischer-zd** [comment]
> Instead of supporting this on per resource basis, it might be easier to set the intended umask value using File#umask() from a recipe or library .. you can also use event handlers to do it before actual convergence begins.

How would this proposal allow us to selectively manipulate file creation nodes?

**Comment by johntdyer** [comment]
@lamont-granquist 

> We do support a config.rb setting for umask so that in 077 environments chef-client itself can be switched to run with 022 which will fix all these problems (maybe violating the letter of the 077 law, but that law is intended to apply to user-created-files by default and these kinds of exceptions with system processes are going to be common).

So it seems this config is only honored when running chef in daemon mode, it seems to be silently ignored otherwise.

**Comment by kalebwalton** [comment]
Any updates on this? This has bitten me so many times. We also run 077 for compliance.

**Comment by kalebwalton** [comment]
Been a few months - any updates or workarounds? I cannot figure out how to have Chef override the umask on a per chef-client run basis and have it do the right thing to directories.

**Comment by lamont-granquist** [comment]
@johntdyer non-daemonized umask was fixed in #6745
